iml-agent/src/daemon_plugins/snapshot.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2020 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+//! Snapshot listing plugin
+//!
+//! This module is responsible for continually discovering snapshots of filesystems.
+//!
+//!
+
+use crate::{
+    action_plugins::lustre::snapshot,
+    agent_error::ImlAgentError,
+    daemon_plugins::{DaemonPlugin, Output},
+    lustre::lctl,
+};
+use async_trait::async_trait;
+use futures::{
+    future::{join_all, AbortHandle, Abortable},
+    lock::Mutex,
+    Future, FutureExt,
+};
+use iml_wire_types::snapshot::{List, Snapshot};
+use std::collections::BTreeSet;
+use std::{pin::Pin, sync::Arc, time::Duration};
+use tokio::time::delay_for;
+
+struct State {
+    updated: bool,
+    output: Output,
+}
+
+#[derive(Debug)]
+pub(crate) struct SnapshotList {
+    reader: Option<AbortHandle>,
+    state: Arc<Mutex<State>>,
+}
+
+pub(crate) fn create() -> SnapshotList {
+    SnapshotList {
+        reader: None,
+        state: Arc::new(Mutex::new(State {
+            updated: false,
+            output: None,
+        })),
+    }
+}
+
+async fn list() -> Result<Vec<Snapshot>, ()> {
+    let fss: Vec<String> = lctl(vec!["get_param", "-N", "mgs.MGS.live.*"])
+        .await
+        .map_err(|e| {
+            tracing::warn!("listing filesystems failed: {}", e);
+        })
+        .map(|o| {
+            o.lines()
+                .map(|line| line.split('.').nth(3).unwrap().to_string())
```

Comment:
perhaps a filter_map instead of an unwrap here.

Reply:
It's warranted to be a `Some` because of `mgs.MGS.live.*`.

Reply:
unwrap / expect is usually a code smell IMO. If we can find a way to to avoid it via a combinator I think it's generally better to do so. 

Justifications that currently hold as an invariant may change in the future and may end up breaking thing unexpectedly.

Reply:
That assumes a parsing error here is something we (and production users) want to know about, not sure that is the case.